rfd/0178-app-custom-icons.md

```diff
@@ -0,0 +1,97 @@
+state: draft
+---
+
+# RFD 178 - Application Resource Custom Icon
+
+## Required Approvers
+
+- Engineering: @romant && @avatus && (someone from the scale team))
+
+## What
+
+Support custom icon uploads for resources through yaml. We can start with feature request [#4712](https://github.com/gravitational/teleport/issues/4712) where
+users provide a URL and Teleport's auth server will download the image and store it in the backend.
+
+## Why
+
+Users want to be able to recognize their applications by famaliar icons/logos rather than reading the application name.
+
+Today, the web UI has some hard coded methods to [guess](https://github.com/gravitational/teleport/blob/dc1199636558f85ed275512175a12207a8034d44/web/packages/shared/components/UnifiedResources/shared/viewItemsFactory.ts#L184) which icon to show for an application by looking at the application name (which may or may not be an accurate depiction of the application) and if nothing matches, we use a generic icon to mean `application`. The guess type is limited to a few handpicked applications.
+
+## Details
+
+### User Story
+
+Alice is a system admin and wants to add custom icon for application resources. Alice specifies the icon URL field on the application resource yaml and using `tctl` creates or updates this file.
+
+### Requirements
+
+- Persist icons in a storage
+- Efficiently retrieve icons
+- Upload icons via yaml resource
+
+### Implementation
+
+#### Persist icons in a storage
+
+Icons can be stored in the backend as blobs with its own row with the key formatted as `icon/<resource-kind>/<name-of-resource>:<uid>`. A unique identifier is required for the proxy cache, otherwise updated icons will not be reflected right away. The UID will be saved in the resource's spec.
+
+#### Efficiently retrieve icons
+
+Currently, in the web UI we fetch 48 resources per listing, that's 48 potential fetch attempts to get the image from the backend from one client. To reduce the amount of trips to the backend, the proxy server will cache the fetched icons in an in-memory sync map that stores icons as byte array with `lastTouched` date to help with cache cleanup.
+
+The maps key will be similar to backend key `<resource-kind>/<name-of-resource>:<uid>`.
+
+Before adding a new icon to cache, we can delete an entry with prefix of `<resource-kind>/<name-of-resource>`, since there will only ever be one icon per resource.
+
+To clean up after possibly deleted icons (eg: deleting an app resource), we can spin a go routine where every day we go through all the values in the map and delete any entries with `lastTouched` date greater than a day.
+
+#### Upload via yaml resource
+
+Users can use the resource spec field `icon` to specify the filepath which Teleport can download the file from.
+
+```proto
+message Icon {
+  // Start with https://
+  string FilePath = 1;
+  // This is an internal field, where we set it
+  // after we successfully uploaded the icon.
+  string UID = 2;
+}
+```
+
+### Security
```

Comment:
**Regarding SSRF**

Assumptions:  i'm changing my implementation to your suggestion where we don't store the image in a backend at all

I thought it was enough to do this (starting when cache is empty):
1. Web UI loads a list of applications, for each application resource that contains a `URL`, send an authenticated request to get this icon
2. Proxy makes the GET request to provided URL (basic URL sanitation and accept only https)
3. Check http response status === 200 OK
2. Do content type check from response header (that it's set to `image/<supported image types>`)
4. Read body up to 101 kb's, check it's under or equal to 100kb
5. Feed this read data into golang's [http#DetectContentType](https://pkg.go.dev/net/http#DetectContentType) just to double check content type
6. Store this data in the proxy cache map where key is the `application/<app-name>:revision`

I may be completely off the mark... but i thought this was enough because:

I understand that images can “look” like images ([fooling](https://dev.to/sebastianstamm/this-image-is-also-a-valid-javascript-file-5fol
) content-type checks to pass) that can also include some code that can be executable. I understood malicious code in the images to be inert, unless the code in the image is processed. In our case, the only processing I can see is:
- In the server: reading the first 512 bytes to determine the content type with go's [http#DetectContentType](https://pkg.go.dev/net/http#DetectContentType)
- in the browser: The browser loading the image through `<img src=”blob://” />`

From what I can scour through internet, my understanding is, as long as our response headers set the correct content type (eg: image/png | jpeg | svg+sml | gif), the browser will not execute any code in the image

Sources:
[img src browser standard](https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-src): here is an [experiment](https://dev.to/sebastianstamm/this-image-is-also-a-valid-javascript-file-5fol
) done to prove it
[SVG](https://www.w3.org/wiki/SVG_Security): about when used as image, JS is not run (since svg's can contain <script> tags)

So if an angry admin decides to idk, provide a URL with an image that has some kind of malicious code, I didn't see the harm as long as it's not processed in other ways than mentioned ^.

Reply:
> Hmm, this seems a little scary from a SSRF perspective, no? Allowing users the ability to make Teleport connect to any arbitrary URL could divulge information about the network or be used as a vector to introduce malware.

@zmb3 also mentioned the ^ and i couldn't come up with an answer for this... 

The other safer alternative is to perhaps require and white list supported cloud storages instead of allowing all random domains? eg. public s3 bucket, oktacdn (this one is more internal, when syncing okta applications to teleport applications)